logs/ldap.rst

```diff
@@ -0,0 +1,90 @@
+============================
+ldap.log and ldap_search.log
+============================
+
+The Lightweight Directory Access Protocol (LDAP) log, or :file:`ldap.log`, is a widely observed protocol
+commonly used for authenticating, directory lookups, centralizing organisational information and accessing client information on email servers. Accordingly, the protocol attracts significant attention from those with adversarial intention.
+
+The LDAP protocol is analyzed using `Spicy <https://docs.zeek.org/projects/spicy/en/latest/>`_. From Zeek version 6.1 onwards the LDAP analyzer is included by default.
+
+LDAP Protocol Overview
+======================
+
+LDAP communicates using a client-server model. The LDAP server contains the directory information and the LDAP client performs operations against this information. This is a quick overview of how the protocol works:
+
+    Sessions: An LDAP session begins with a client connecting to an LDAP server, optionally securing the connection with encryption, and then binding to the server by providing credentials.
+
+    Queries: Clients search for entries in the LDAP directory using LDAP queries, which consist of a base DN, a scope (such as one level or the entire subtree), and a filter to match entries. Queries are read only.
+
+    Operations: Clients with the correct privileges can perform a variety of operations; in addition to search, they can add, delete or modify.
+
+    Data Format: LDAP data entries are formatted as records consisting of a distinguished name (DN) and a set of attributes. Each attribute has a name and one or more values.
+
+The LDAP analyzer outputs two LDAP related logs. :file:`ldap.log` contains details about the LDAP session except those related to searches. :file:`ldap_search.log` contains information related to LDAP searches.
+
+For details on every element of the :file:`ldap.log` and :file:`ldap_search.log` refer to :zeek:see:`LDAP`. Below is an inspection of the :file:`ldap.log` and :file:`ldap_search.log` in JSON format.
+
+ldap.log
+========
+
+An example of an :file:`ldap.log`.
+
+.. code-block:: console
+
+    zeek@zeek-6.1:~ zeek -C LogAscii::use_json=T LDAP::default_log_search_attributes=T -r ldap-simpleauth.pcap
+    zeek@zeek-6.1:~ jq . ldap_search.log
+
+::
+
+    {
+      "ts": 1463256456.051759,
+      "uid": "ChD43F3guxAmJ5f2aj",
+      "id.orig_h": "10.0.0.1",
+      "id.orig_p": 25936,
+      "id.resp_h": "10.0.0.2",
+      "id.resp_p": 3268,
+      "message_id": 3,
+      "version": 3,
+      "opcodes": ["bind simple"],
+      "results": ["success"],
+      "objects": ["CN=xxxxxxxx,OU=Users,OU=Accounts,DC=xx,DC=xxx,DC=xxxxx,DC=net"],
+      "arguments": ["REDACTED"]
```

Comment:
@mbispham - do you have insight (or access to production data logs) whether using sets/vectors for these fields is reasonable? Could these be scalar values instead. If there's a few special cases here and and there, we could consider unrolling them instead.

Main thought is ingestion/filtering of multi-value fields that may be annoying for downstream consumers when in practice there's only a single value (same for ldap_search.log).

Thanks for the PR! Much appreciated!

(cc @mmguero)

Reply:
This [aggregates state across multiple messages](https://github.com/zeek/zeek/blob/4eb1b71d1b29845dec216536d830e79f2f95959b/scripts/base/protocols/ldap/main.zeek#L107-L110), so I think unless we want to new log syntax this should be a container.

These fields already contain [string representations of more complicated LDAP constructs](https://github.com/zeek/zeek/blob/4eb1b71d1b29845dec216536d830e79f2f95959b/testing/btest/Baseline/scripts.base.protocols.ldap.attributes/ldap_search.log#L10) which I believe were modeled after what Wireshark does. We should check whether Wireshark does anything interesting for lists (though I doubt that since they often do not aggregate such state across packages), but users seem to expect Wireshark format, https://github.com/zeek/spicy-ldap/issues/21, https://github.com/zeek/spicy-ldap/issues/23, https://github.com/zeek/spicy-ldap/issues/27.

Reply:
> This [aggregates state across multiple messages](https://github.com/zeek/zeek/blob/4eb1b71d1b29845dec216536d830e79f2f95959b/scripts/base/protocols/ldap/main.zeek#L107-L110), so I think unless we want to new log syntax this should be a container.

That table is used to correlate requests and responses based on message_id. A server may reply to requests out-of-order so that makes sense.

Using vectors for opcode, result and all the other fields is a different question: If a request/response pair is only ever about a single opcode (or an error), then these shouldn't be modeled as vectors. With the exception of "search" (which is treated separately) the RFC indicates this is the case.

I'm mostly just sensitive as it's very related to the tx_hosts / rx_hosts fields that were removed from the files.log: In practice, they only ever contained a single entry, but being represented as arrays they needed special treatment downstream.

Reply:
@awelzel - I think this is a worthwhile discussion - but not for this PR.